src/Types/ContractAddress.cs

```diff
@@ -0,0 +1,77 @@
+using Concordium.Sdk.Helpers;
+
+namespace Concordium.Sdk.Types;
+
+/// <summary>
+/// Represents a contract address.
+///
+/// A contract is identified by its unique <see cref="ContractAddress"/>.
+/// A contract address consists of a contract index and a contact sub-index
+/// both represented as 64-bit unsigned integers.
+/// </summary>
+public readonly struct ContractAddress : IEquatable<ContractAddress>
+{
+    /// <summary>
+    /// The index part of the address of a contract.
+    /// </summary>
+    public readonly ulong Index { get; init; }
+
+    /// <summary>
+    /// The sub-index part of the address of a contract.
+    /// </summary>
+    public readonly ulong SubIndex { get; init; }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ContractAddress"/> class.
+    /// </summary>
+    /// <param name="index">The contract index.</param>
+    /// <param name="subIndex">The contract sub index.</param>
+    private ContractAddress(ulong index, ulong subIndex)
+    {
+        this.Index = index;
+        this.SubIndex = subIndex;
+    }
+
+    /// <summary>
+    /// Creates an instance of contract address.
+    /// </summary>
+    /// <param name="index">the index value.</param>
+    /// <param name="subIndex">the sub index value.</param>
+    public static ContractAddress Create(ulong index, ulong subIndex) => new(index, subIndex);
+
+    /// <summary>
+    /// Converts the contract address to its corresponding protocol buffer message instance.
+    ///
+    /// This can be used as the input for class methods of <see cref="Client.RawClient"/>.
+    /// </summary>
+    public Grpc.V2.ContractAddress ToProto() => new() { Index = Index, Subindex = SubIndex };
+
+    public bool Equals(ContractAddress other) => this.Index == other.Index && this.SubIndex == other.SubIndex;
+
+    public override bool Equals(object? obj)
+    {
+        if (obj is null)
+        {
+            return false;
+        }
+
+        if (obj.GetType() != this.GetType())
+        {
+            return false;
+        }
+
+        var other = (ContractAddress)obj;
+        return this.Equals(other);
+    }
+
+    public override int GetHashCode()
+    {
+        var indexBytes = Serialization.GetBytes(this.Index);
```

Comment:
Since you are using the hash of a array you don’t get the same hash - you don’t even get the same hash on the same object. This because the hash is taking on the reference.

Try run this test
```
    [Fact]
    public void SameHash()
    {
        // Arrange
        var first = ContractAddress.Create(1, 1);
        var second = ContractAddress.Create(1, 1);

        // Act
        var firstHash = first.GetHashCode();
        var secondHash = second.GetHashCode();
        var firstHashAgain = first.GetHashCode();

        // Assert
        firstHash.Should().Be(firstHashAgain);
        firstHash.Should().Be(secondHash);
    }
```

you will see that none of the asserts succeeds.